org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/util/CodeCompletionExecutor.java

```diff
@@ -0,0 +1,98 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Red Hat - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.internal.corext.util;
+
+import java.util.HashSet;
+
+import org.eclipse.core.runtime.NullProgressMonitor;
+
+import org.eclipse.jdt.core.CompletionProposal;
+import org.eclipse.jdt.core.CompletionRequestor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.Signature;
+
+/**
+ * Copied from {@link JavaModelUtil#getStaticImportFavorites(ICompilationUnit, String, boolean, String[])}
+ * If code completion needs to be called multiple times on a given compilation unit,
+ * this class will perform the operations more efficiently by re-using a working copy.
+ */
+public class CodeCompletionExecutor {
```

Comment:
Maybe I could call this something better. CodeCompletionInvoker ? Also, given that this is copied from `JavaModelUtil` I should keep the original license header. Although various code blocks were relocated.